### PR TITLE
Issue #7671: Update doc for NoFinalizer

### PR DESCRIPTION
**Pull request by OmarShawky1**
Issue #7671: added coding Example.

image

$ cat Test.java
public class Test {

    protected void finalize() throws Throwable { // violation
        try {
            System.out.println("overriding finalize()");
        } catch (Throwable t) {
            throw t;
        } finally {
            super.finalize();
        }
    }
}

$ java -jar checkstyle-8.31-all.jar -c /google_checks.xml Test.java
Starting audit...
[WARN] /mnt/c/Users/Ipowe/Desktop/Test.java:3:5: Avoid using finalizer method. [NoFinalizer]
Audit done.

### CODE REVIEW

**Comment by romani** [comment]
@OmarShawky1 , please resolve conflicts and strictly follow https://github.com/checkstyle/checkstyle/wiki/PR-rules , we will not review PRs that are messy

**Comment by OmarShawky1** [comment]
i need help, i do not know what is exactly the error

**Comment by romani** [comment]
on your local, please back backup of your changes, drop your branch, get latest from our common repo - https://checkstyle.org/beginning_development.html
```
git pull main master
git push origin master
git checkout -b your-branch
# apply all changes again
git add .
git commit -m "........"
git push origin newPR -f
```

**Comment by OmarShawky1** [comment]
the error says that "gc" is a grammar error while it is garbage collector and it is not an error, should i remove it?

**Comment by romani** [comment]
@OmarShawky1, please read attentively all in issue description, and all links referenced, please look at similar already merged PRs and try to figure out what is wrong with your PR description.
We will not start review until  PR is properly show all details.

**Comment by OmarShawky1** [comment]
waiting review

**Comment by strkkk** [comment]
@OmarShawky1 before review, PR should be formatted to follow rules.
Also, it is better to keep examples in same format, please see merged doc PRs for examples.

**Comment by rdiachenko** [comment]
@OmarShawky1 what's the status?

**Comment by OmarShawky1** [comment]
my apologies for the late response, i did not understand what you are referring to exactly

**Comment by OmarShawky1** [comment]
i looked into pull requests that has approved reviews and i did not quite understand what i'm missing?
for example this pull requests look similar to me regarding the PR description:
https://github.com/checkstyle/checkstyle/pull/7988

and the following PR elaborates a lot in the PR description
https://github.com/checkstyle/checkstyle/pull/8017

so excuse me if i did not quite get what is missing/incorrect with my PR description

**Comment by wltan** [comment]
The admins want you to follow https://github.com/checkstyle/checkstyle/wiki/Good-practice-on-how-to-update-examples, which is also referenced in #7671 itself. Specifically, the last 3 paragraphs:

> Please share in PullRequest (PR) screenshot (drag and drop image to PR description) of how website looks like after your update. You can generate web site by command https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations#how-to-generate-website-only
> 
> Please share prove that your samples works in our latest release as you describe in documentation. Please use our CLI tool and share output on all new samples in format of https://github.com/checkstyle/checkstyle/blob/master/.github/ISSUE_TEMPLATE.md (only terminal output)
> 
> Good example of how PullRequest (PR) should looks like - https://github.com/checkstyle/checkstyle/pull/7634 .

**Comment by OmarShawky1** [comment]
awaiting review to accept merge

**Comment by rdiachenko** [comment]
@OmarShawky1 
> $ java -jar checkstyle-8.31-all.jar -c/google_checks.xml Test.java
Starting audit...
[WARN] /home/omar/Desktop/Test.java:18:5: Top-level class NoFinalizerClass has to reside in its own source file. [OneTopLevelClass]
Audit done.

Please update your testing example to give a correct violation for `NoFinalizer` check. The violation from your example belongs to a different check.

**Comment by OmarShawky1** [comment]
awaiting review

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java [hunk 1]
```diff
@@ -45,6 +45,22 @@
  * <pre>
  * &lt;module name=&quot;NoFinalizer&quot;/&gt;
  * </pre>
+ * <p>Example:</p>
+ * <pre>
+ *  public class Test {
+ *
+ *      // Violation
+ *      protected void finalize() throws Throwable {
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:53]: Please add a second `finalize` method but with parameters, there should not be a violation for it.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java [hunk 2]
```diff
@@ -45,6 +45,22 @@
  * <pre>
  * &lt;module name=&quot;NoFinalizer&quot;/&gt;
  * </pre>
+ * <p>Example:</p>
+ * <pre>
+ *  public class Test {
+ *
+ *      // Violation
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:52]: Please put `// Violation` at the end of line where the actual violation occurs.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java [hunk 3]
```diff
@@ -45,6 +45,22 @@
...
+ *      protected void finalize() throws Throwable {
+ *          try {
+ *          System.out.println("overriding finalize()");
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:55]: Missed few whitespaces, please format it correctly.

File: src/xdocs/config_coding.xml [hunk 4]
```diff
@@ -4468,6 +4468,24 @@ enum Foo10 { FOO, BAR } // OK
         <source>
 &lt;module name=&quot;NoFinalizer&quot;/&gt;
         </source>
+        <p>Example:</p>
+          <div class="wrapper">
+              <pre>
```
Comment (rdiachenko) [inline src/xdocs/config_coding.xml:4473]: Don't use these tags here. Please take a look at other pull requests and update it accordingly.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java (hunk 2 above)
Comment (OmarShawky1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:52]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java (hunk 3 above)
Comment (OmarShawky1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:55]: done

File: src/xdocs/config_coding.xml (hunk 4 above)
Comment (OmarShawky1) [reply src/xdocs/config_coding.xml:4473]: done; i removed only div, i wrote div as it was mentioned in the example only.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java (hunk 1 above)
Comment (OmarShawky1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:53]: done

File: src/xdocs/config_coding.xml (hunk 4 above)
Comment (OmarShawky1) [reply src/xdocs/config_coding.xml:4473]: tried to remove div & pre, it gave me an error (as expected), tried to remove div only, i got an error:
image

Comment (rdiachenko) [reply src/xdocs/config_coding.xml:4473]: `<source>` tag should be used instead

Comment (OmarShawky1) [reply src/xdocs/config_coding.xml]: done

**Comment by OmarShawky1** [comment]
i couldn't understand the error:
https://travis-ci.org/github/checkstyle/checkstyle/jobs/677079945#L2126
https://travis-ci.org/github/checkstyle/checkstyle/jobs/677079945#L2132
what did i miss?

**Comment by rdiachenko** [comment]
> i couldn't understand the error:
> https://travis-ci.org/github/checkstyle/checkstyle/jobs/677079945#L2126
> https://travis-ci.org/github/checkstyle/checkstyle/jobs/677079945#L2132
> what did i miss?

Please rebase on latest master

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java [hunk 5]
```diff
@@ -45,6 +45,31 @@
  * <pre>
  * &lt;module name=&quot;NoFinalizer&quot;/&gt;
  * </pre>
+ * <p>Example:</p>
+ * <pre>
+ *  public class Test {
+ *
+ *      protected void finalize() throws Throwable { //Violation
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:52]: Please change `//Violation` to `// violation`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java [hunk 6]
```diff
@@ -45,6 +45,31 @@
...
+ *          try {
+ *              System.out.println("overriding finalize()");
+ *          } catch (Throwable t) {
+ *              throw t;
+ *          } finally {
+ *              super.finalize();
+ *          }
+ *      }
+ *
+ *      protected void finalize(Object object) throws Throwable{
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:62]: `Throwable{` -> `Throwable { // OK, overloaded method`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java [hunk 7]
```diff
@@ -45,6 +45,31 @@
...
+ *          try {
+ *              System.out.println("overriding finalizer() having parameters");
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:64]: Remove one whitespace before `System...` to be in line with other statements

File: src/xdocs/config_coding.xml [hunk 8]
```diff
@@ -4468,6 +4468,31 @@ enum Foo10 { FOO, BAR } // OK
         <source>
 &lt;module name=&quot;NoFinalizer&quot;/&gt;
         </source>
+        <p>Example:</p>
+          <source>
```
Comment (rdiachenko) [inline src/xdocs/config_coding.xml:4472]: Please fix indentation. Should be the same level as `<p>`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java (hunk 5 above)
Comment (OmarShawky1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:52]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java (hunk 6 above)
Comment (OmarShawky1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:62]: done

File: src/xdocs/config_coding.xml (hunk 8 above)
Comment (OmarShawky1) [reply src/xdocs/config_coding.xml:4472]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java (hunk 7 above)
Comment (OmarShawky1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:64]: should i apply this to all the the try and catch and final ?

Comment (rdiachenko) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java]: Yes please. Make it consistent through out your code. I still see one case with this line not properly formatted.

Comment (OmarShawky1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java]: done

**Comment by OmarShawky1** [comment]
couldn't relate to this error:
https://travis-ci.org/github/checkstyle/checkstyle/jobs/678765016#L185

i'm am updated to the latest master on the original repo

File: src/xdocs/config_coding.xml [hunk 9]
```diff
@@ -4468,6 +4468,31 @@ enum Foo10 { FOO, BAR } // OK
...
+        }
+    }
+
+    protected void finalize(Object object) throws Throwable { // OK, overloaded method
+        try {
+           System.out.println("overriding finalizer() having parameters");
+        } catch (Throwable t){
+           throw t;
+        } finally {
+           super.finalize();
```
Comment (pbludov) [inline src/xdocs/config_coding.xml:4491]: The `object` parameter is missing here.

Comment (rdiachenko) [reply src/xdocs/config_coding.xml:4491]: I think it is fine, it is a call to original `finalize()`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java [hunk 10]
```diff
@@ -45,6 +45,31 @@
...
+ *             System.out.println("overriding finalize()");
+ *          } catch (Throwable t) {
+ *             throw t;
+ *          } finally {
+ *             super.finalize();
+ *          }
+ *      }
+ *
+ *      protected void finalize(Object object) throws Throwable { // OK, overloaded method
+ *          try {
```
Comment (rdiachenko) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/NoFinalizerCheck.java:63]: remove one whitespace to correct indentation

File: src/xdocs/config_coding.xml (hunk 9 above)
Comment (OmarShawky1) [reply src/xdocs/config_coding.xml:4491]: so, should i add it? and if yes, why?

Comment (pbludov) [reply src/xdocs/config_coding.xml:4491]: > I think it is fine, it is a call to original `finalize()`

A call to original `finalize()`? A quote from the Check description: `Finalizers are unpredictable, often dangerous, and generally unnecessary. Their use can cause erratic behavior, poor performance, and portability problems.`
The example should do not call the original `finalize`. And `finalize(object)` is not a finalizer. It is just a method with a poorly chosen name. It has nothing to do with GC.

Comment (OmarShawky1) [reply src/xdocs/config_coding.xml:4491]: image
i think this is a syntax error as this class is not inheriting 'finalizer(Object object)'

Comment (rdiachenko) [reply src/xdocs/config_coding.xml:4490]: > The example should do not call the original finalize

@pbludov , could you explain why? I thought the purpose of the whole `NoFinalizerCheck` check is to complain if you override the original `finalize` method (the first method above) but in the second method above you just call the original `finalize`. So, why there should not be a call to original `finalize` from another method?

Comment (pbludov) [reply src/xdocs/config_coding.xml:4490]: > i think this is a syntax error as this class is not inheriting 'finalizer(Object object)'

Why? `finalize` is not a reserved word. The example shows just a method with name `finalize`. This method is not called by GC.

**Maybe** the OK example should be just dropped, as it is already confusing people.
@rdiachenko, what are you think? The Check is not expected to have a **valid** case.

Comment (rdiachenko) [reply src/xdocs/config_coding.xml:4490]: If it confuses people I'm fine to remove this case. @OmarShawky1 could you please leave a `finalize` method with violation only?

Comment (OmarShawky1) [reply src/xdocs/config_coding.xml]: done

File: src/xdocs/config_coding.xml [hunk 11]
```diff
@@ -4468,6 +4468,21 @@ enum Foo10 { FOO, BAR } // OK
...
+        <source>
+public class Test {
+
+    protected void finalize() throws Throwable { // violation
+        try {
+           System.out.println("overriding finalize()");
+        } catch (Throwable t) {
+           throw t;
+        } finally {
+           super.finalize(object);
```
Comment (rdiachenko) [inline src/xdocs/config_coding.xml:4481]: Please remove `object` argument. The code doesn't compile with it.

Comment (OmarShawky1) [reply src/xdocs/config_coding.xml:4481]: done

**Pull request merged**